File: bots/rs/v55/src/builder/tasks/_policy.rs
//! Translation of `bots/intgrah/v54.7.9/builder/tasks/_policy.py`.
//!
//! Tree-structured policy primitives.
//!
//! A `Policy` is either a `TaskGroup` (an internal node with named children
//! and an optional gate) or a leaf function `LeafFn` of shape
//! `(Builder, Controller) -> TaskResult`. Leaves either complete the turn
//! (return `Ok(())`) or return `Err(TaskRejected)` to defer to the next
//! sibling.
//!
//! Traversal: depth-first, first-success-wins. `run_policy` returns true
//! iff some leaf in the subtree completed without rejecting; false iff
//! every leaf rejected (or the group's gate denied the subtree). The
//! caller's parent group treats a false return the same way it treats a
//! leaf rejection — move on to the next sibling.

use cambc::Controller;

use crate::builder::Builder;
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};
use crate::util::debug::{Scope, debug as log};

use serde_json::Map;

pub type LeafFn = fn(&mut Builder, &mut Controller<'_>) -> TaskResult;
pub type Gate = fn(&mut Builder, &mut Controller<'_>) -> bool;

/// Internal policy node. `children` is searched in order; `gate`, if set,
/// can short-circuit the entire subtree when its precondition doesn't
/// hold (cheaper than rejecting at every leaf separately).
pub struct TaskGroup {
    pub name: &'static str,
    pub children: &'static [Policy],
    pub gate: Option<Gate>,
}

#[derive(Clone, Copy)]
pub enum Policy {
    Group(&'static TaskGroup),
    Leaf { name: &'static str, fn_: LeafFn },
}

pub fn run_policy(self_: &mut Builder, ct: &mut Controller<'_>, policy: &Policy) -> bool {
    match policy {
        Policy::Group(group) => {
            if let Some(gate) = group.gate
                && !gate(self_, ct)
            {
                let mut args = Map::new();
                args.insert(
                    "name".to_string(),
                    serde_json::Value::String(group.name.to_string()),
                );
                log("{name}: gated off", args);
                return false;
            }
            let _scope = Scope::new_timed(group.name);
            for child in group.children {
                if run_policy(self_, ct, child) {
                    return true;
                }
            }
            false
        }
        Policy::Leaf { name, fn_ } => {
            let scope_label = format!("task={name}");
            let _scope = Scope::new_timed(&scope_label);
            match fn_(self_, ct) {
                Ok(()) => true,
                Err(rej) => {
                    let mut args = Map::new();
                    args.insert(
                        "name".to_string(),
                        serde_json::Value::String((*name).to_string()),
                    );
                    args.insert(
                        "reason".to_string(),
                        serde_json::Value::String(rej.reason.clone()),
                    );
                    log("{name}: {reason}", args);
                    false
                }
            }
        }
    }
}

File: bots/beacon/bestrn/builder/tasks/_policy.py
"""
Translation of `bots/intgrah/v54.7.9/builder/tasks/_policy.py`.

Tree-structured policy primitives.

A `Policy` is either a `TaskGroup` (an internal node with named children
and an optional gate) or a leaf function `LeafFn` of shape
`(Builder, Controller) -> TaskResult`. Leaves either complete the turn
(return `None`) or return `Err(TaskRejected)` to defer to the next
sibling.

Traversal: depth-first, first-success-wins. `run_policy` returns true
iff some leaf in the subtree completed without rejecting; false iff
every leaf rejected (or the group's gate denied the subtree). The
caller's parent group treats a false return the same way it treats a
leaf rejection — move on to the next sibling.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cambc import Controller
if TYPE_CHECKING:
    from builder import Builder
if TYPE_CHECKING:
    from builder.tasks.rejected import TaskResult
from util.debug import Scope
from util.debug import debug as log

type LeafFn = Callable[[Builder, Controller], TaskResult]
type Gate = Callable[[Builder, Controller], bool]


class TaskGroup:
    """
    Internal policy node. `children` is searched in order; `gate`, if set,
    can short-circuit the entire subtree when its precondition doesn't
    hold (cheaper than rejecting at every leaf separately).
    """

    name: str
    children: list[Policy]
    gate: Gate | None

    def __init__(self, name: str, children: list[Policy], gate: Gate | None) -> None:
        self.name = name
        self.children = children
        self.gate = gate


@dataclass(frozen=True, slots=True)
class PolicyGroup:
    _0: TaskGroup


@dataclass(frozen=True, slots=True)
class PolicyLeaf:
    name: str
    fn_: LeafFn


type Policy = PolicyGroup | PolicyLeaf


def run_policy(self_, ct, policy) -> bool | None:
    match policy:
        case PolicyGroup(_0=group):
            gate = group.gate
            if gate is not None and (not gate(self_, ct)):
                args = {}
                args["name"] = str(group.name)
                log("{name}: gated off", args)
                return False
            with Scope.new_timed(group.name) as _scope:
                return any(run_policy(self_, ct, child) for child in group.children)
        case PolicyLeaf(name=name, fn_=fn_):
            scope_label = f"task={name}"
            with Scope.new_timed(scope_label) as _scope:
                match fn_(self_, ct):
                    case None:
                        return True
                    case rej if rej is not None:
                        args = {}
                        args["name"] = str(name)
                        args["reason"] = rej.reason
                        log("{name}: {reason}", args)
                        return False
